File: web/src/application/use-cases/send-contact-message.ts
import { type Failure, isFailure } from "@domain/failures/failure";
import type { ContactMessageRepository } from "@domain/repositories/types";
import { type ContactMessage, parseContactMessage } from "@domain/value-objects/contact-message";

type DeliverContactMessage = ContactMessageRepository["deliver"];

export interface SendContactMessageParams {
	name?: string | null;
	email: string;
	body: string;
}

export const sendContactMessage =
	(deliver: DeliverContactMessage) =>
	async ({ name, email, body }: SendContactMessageParams): Promise<ContactMessage | Failure> => {
		const message = parseContactMessage({ name, email, body });
		if (isFailure(message)) return message;

		return deliver(message);
	};

File: app/lib/application/use_cases/send_contact_message.dart
import 'package:contribkit/domain/repositories/contact_message_repository.dart';
import 'package:contribkit/domain/value_objects/contact_message.dart';

final class SendContactMessage {
  const SendContactMessage({required this.repository});

  final ContactMessageRepository repository;

  Future<void> call(ContactMessage message) => repository.deliver(message);
}
